### ISSUE TITLE
Instantiation of opennebula_virtual_machine with just template_id does not pass template attributes

### ISSUE THREAD

**Issue by noroute** [OpenNebula/addon-terraform#1]
Template settings (e.g. MEMORY) do not seem to be passed on to the OpenNebula API when instantiating a template. Instantiation uses the right `template_id`, the template is also correct (calling `onetemplate instantiate` via CLI works).

When I add `memory` as attribute for the `opennebula_virtual_machine` directly, it is passed on correctly (but then complains about the next missing attribute).

Am I missing something here?

```
resource "opennebula_virtual_machine" "foo" {
  template_id = "2553"
}
```

`terraform apply`

```bash  
# opennebula_virtual_machine.foo will be created
  + resource "opennebula_virtual_machine" "foo" {
      + gname       = (known after apply)
      + id          = (known after apply)
      + instance    = (known after apply)
      + ip          = (known after apply)
      + lcmstate    = (known after apply)
      + pending     = false
      + permissions = (known after apply)
      + state       = (known after apply)
      + template_id = 2553
      + uid         = (known after apply)
      + uname       = (known after apply)
    }

Plan: 1 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

opennebula_virtual_machine.foo: Creating...
opennebula_template.foo: Modifying... [id=2556]
opennebula_template.foo: Modifications complete after 0s [id=2556]

Error: OpenNebula error [ALLOCATE]: [one.template.instantiate] Error allocating a new virtual machine template. MEMORY attribute must be a positive integer value.

  on main.tf line 15, in resource "opennebula_virtual_machine" "foo":
  15: resource "opennebula_virtual_machine" "foo" {
```

`onetemplate show 2553`

```
TEMPLATE 2553 INFORMATION
ID             : 2553
NAME           : FOO
USER           : xxx
GROUP          : xxx
LOCK           : None
REGISTER TIME  : xxx

PERMISSIONS
OWNER          : um-
GROUP          : ---
OTHER          : ---

TEMPLATE CONTENTS
CONTEXT=[
  HOSTNAME="$name",
  NETWORK="YES" ]
CPU="0.2"
DISK=[
  IMAGE="xxx" ]
FEATURES=[
  ACPI="yes" ]
MEMORY="2048"
OS=[
  ARCH="x86_64",
  BOOT="disk0" ]
VCPU="2"
```

**Comment by jaypif** [comment]
Hi @noroute ,

Actually what you are doing is logical but for the moment you must add `memory = 2048`parameter in addition to the `template_id`.
The issue is not on the Terraform part it is directly on OpenNebula API.
`one.template.instantiate` returns the error you faced if the memory is not set whereas it is already the case in your template.
I faced the same issue like you during my test.

It is something we want to improve within OpenNebula Team thanks to this provider.

So to fix your issue, as a workaround, you can override the memory value in the resource parameter:
```
resource "opennebula_virtual_machine" "foo" {
  template_id = "2553"
  memory = 2048
}
```

Regards

**Comment by noroute** [comment]
Hi @jaypif,

thanks for the quick answer. This workaround sounds a little clumsy as it will force me to set all the attributes (cpu, disk, etc.) on every virtual machine, doesn't it?

How does the `onetemplate instantiate` CLI do it? Doesn't it use the same API? Or is it a `goca` issue?

**Comment by christian7007** [comment]
I think the problem is that [here](https://github.com/OpenNebula/addon-terraform/blob/62f6448633e638f0877a004f58795cdc7664f06d/opennebula/resource_opennebula_virtual_machine.go#L814) the memory value is set to 0 if you don't provide a value. So when you instantiate the template the oned receive an extra template with a memory value set to 0 which overrides the current memory value and makes the instantiation to fail.

**Comment by noroute** [comment]
@christian7007 Without having written tests for that this seems plausible. When creating a new VM, the whole _schema_ contents of `ResourceData` is always included as extra XML, overwriting the data from the template. 

Apparently, the extra XML must only include *non-default values* to allow selective overwriting of template values. `[d.Get` always returns _non-nil default values_ for keys that are included in the schema](https://github.com/hashicorp/terraform/blob/8b2646c2a6676c62a0ca2531d95b392f412b5abc/helper/schema/resource_data.go#L63), even if they are not included in the configuration data. `d.GetOk` seems to be the better way.

Generation of extra XML from ResourceData: https://github.com/OpenNebula/addon-terraform/blob/62f6448633e638f0877a004f58795cdc7664f06d/opennebula/resource_opennebula_virtual_machine.go#L431

Instantiation with extra xml: https://github.com/OpenNebula/addon-terraform/blob/62f6448633e638f0877a004f58795cdc7664f06d/opennebula/resource_opennebula_virtual_machine.go#L438

**Comment by jaypif** [comment]
Hi @christian7007 and @noroute ,

You are right, the XML generation override XML from file. In this case it sets the memory to 0 while it is already present in you XML.
However, as it is possible to override some parameters while there are in the template it is not really simple to satisfy all the possible cases. Actually, the XML structure of the template is that every parameters (mandatory of non mandatory) are part of the `TEMPLATE` part of the XML. 

Here is what I suggest:
- Case 1:
If the user set the `template_id` add only the content of "template" as extra template.
- Case 2:
If the user create a VM directly (no `template_id`), generate XML for the VM directly.

I think this will simplify the behavior. The `resource_opennebula_template` can be used before to create a template, get its `id` and re-use this `id` as `template_id`.

Does it seems correct for you ?

**Comment by noroute** [comment]
@christian7007 This would be one simple way to do it. Note that this prevents users from creating VMs from templates with customizations, which is a limitation.

What (besides needing good test coverage) prevents you from only adding attributes the extra XML that are actually present (the Terraform schema helper has `d.GetOk` to detect cases of missing configuration)?

**Comment by jaypif** [comment]
@noroute 

The case 1 I suggest will load what is srt as `template` as extra template during instantiation. So if in the template file there are some values for `MEMORY` or any other resources, it will override what is already exists in the template.
In that case, if there is an error, it will be in the file directly.

About the default value issue, you are right using `d.GetOk` is better. Nevertheless, based on my  "Case 2" suggestion, some parameter will be used only during VM creation if the `template_id` is not set. So it means that if there is not `template_id`, if memory is also empty, Terraform  will raise an error.
To prevent this, `ConflictWith` can be use in the Resource descrition.

Regards

**Comment by christian7007** [comment]
Hello @noroute and @jaypif,

I don't know much about Terraform but if the solution provided by @jaypif allow to instantiate VMs from a template either overriding some attributes or not (as he said in hist last comment) I think is ok.

**Comment by jaypif** [comment]
Actually I am not really satisfied with what I suggest.
I would like to be able to instantiate a template and attach NICs or Disk I just create in Terraform in the same time.
With What I suggest it is not possible to override disk or nic, the template mst already have them while in my usage it is not the case.

Example: I have Templates with only CPU , VCPU and RAM fixed (not possible to change them after).
When I instantiate a template, I want to attach NIC and/or disk I create with other terraform resources. 

So I want to propose this new idea:
If `template_id` is set, not possible to use "cpu" or "memory" parameters but still possible to append extra template and to also attach directly to the VM additional parameters such as Disk or Nic.
If the template_id is not set, the VM is created from scratch.

Does it suit better for you ?

**Comment by noroute** [comment]
Retaining the possibility for customizations sounds very good.

But what's the problem with `cpu` and `memory`? Using `d.GetOk` to determine if there actually _is_ and override for these in the Terraform resource definition and _only then_ rendering them into the extra XML sounds sensible, doesn't it? This way, the value from the template would be used _unless_ it is also defined in Terraform.

**Comment by jaypif** [comment]
You are right.

I use `d.GetOk` only if the template_id is not used but not in the XML creation. That was my mistake.
So I will remove the `extra_template` because I think it is better to keep them in the template resource.
So in this case I will just fix the cpu and memory with GetOk in the XML generation.

**Comment by jaypif** [comment]
I just did a PR #2 to fix the wrong behavior.

If it is Ok for you it can be merged

**Comment by noroute** [comment]
So in your solution, `cpu` and `memory` cannot be overridden. I understood your comment in https://github.com/OpenNebula/addon-terraform/issues/1#issuecomment-516348354 as to make overriding possible without breaking the non-override case.

Also, does the project have unit test infrastructure in place to cover the changes you're about to make?

**Comment by jaypif** [comment]
You are right, overriding memory or cpu should be possible. So my PR is not correct, I will change it.
I will remove the constraint and let the different tests with `d.getOk`

**Comment by jaypif** [comment]
The PR has been updated

**Comment by jaypif** [comment]
Hi @noroute 

Is my last version OK for you ? If yes, I will merge it.

About test infrastructure, OpenNebula deployed an infra for that. I added some test already in master (and I have other ones for missing resources available). 
You can ask @christian7007 for credentials if you want to use it for your tests.

Regards

**Comment by noroute** [comment]
@jaypif The code looks like it could work. I still don't see any unit-level tests. Were you planning to add those? Is there an overall strategy for testing `addon-terraform`? I guess adoption and collaboration could benefit from easily accessible (unit-level) tests.

**Comment by jaypif** [comment]
Hi @noroute 

Good!
About test I added them in master (for starting) and I have more availble not pushed yet.
@christian7007 provided an environment available for testing.
A CI can be added too with a docker where opennebula can be setup quickly (thanks to minione), this part is not available yet.

OK So I propose to merge the fix about this bug and we can deal testing discussions on another issue

**Comment by jaypif** [comment]
Hi @noroute 

@christian7007 connected this repo to Travis CI. So now we have CI on all tests and test coverage too thanks to cedecov.io.

The CI is already available, the codecov is under PR: #6 

This issue has been fixed by #2
